internal/core/streams/stream_provider_server_side_test.go

```diff
@@ -202,4 +203,85 @@ func TestStreamProviderServerSide(t *testing.T) {
 			verifyHandlerHeartbeat(t, sp, esp, validCredential)
 		})
 	})
+
+	t.Run("Replay", func(t *testing.T) {
+		store := makeMockStore(nil, nil)
+		store.latency = 100 * time.Millisecond
+
+		repo := &serverSideEnvStreamRepository{store: store, loggers: ldlog.NewDisabledLoggers()}
+
+		newData := []ldstoretypes.Collection{
+			{Kind: ldstoreimpl.Features(), Items: store.flags},
+			{Kind: ldstoreimpl.Segments(), Items: store.segments},
+		}
+		expected := MakeServerSidePutEvent(newData)
+		expectedData := expected.Data()
+		t.Run("ConsecutiveCalls", func(t *testing.T) {
+			eventChannel := repo.Replay("", "")
+			select {
+			case actual := <-eventChannel:
+				if actual == nil {
+					t.Fatal("Expected an event from Replay")
+				}
+				if expectedData != actual.Data() {
+					t.Errorf("Replay() = %v; wanted %v", actual.Data(), expectedData)
+				}
+			case <-time.After(time.Second):
+				t.Error("expected an event to be returned")
+			}
+			eventChannel = repo.Replay("", "")
+			select {
+			case actual := <-eventChannel:
+				if actual == nil {
+					t.Fatal("Expected an event from Replay")
+				}
+				if expectedData != actual.Data() {
+					t.Errorf("Replay() = %v; wanted %v", actual.Data(), expectedData)
+				}
+			case <-time.After(time.Second):
+				t.Error("expected an event to be returned")
+			}
+
+		})
+		t.Run("ConcurrentCalls", func(t *testing.T) {
```

Comment:
As part of the unit test polishing work I'm doing, I'm changing how this one works because I think there's a logic error in it.

The test on line 282 does not, I think, actually verify that the computation was not done twice: `event1 != event2` is comparing two interface values, and in Go that is not a _reference_ equality test (as in, "are they literally the same instance of the type"), it is a test of whether they have the same concrete types and whether their values as those concrete types are equal. In other words, `a` and `b` are separately created instances of some struct type and their fields have the same values, and you cast them both to some interface type, the resulting interface values are equal. And since the data store in this case returns the same flag/segment data every time it is queried (the same _empty_ data, since it was created with flags and segments of `nil`), this test can't possibly fail. So I'm implementing this differently, using a store that is instrumented to return different data for each query.

Reply:
Sorry, my previous comment was not quite right. The test as written does work, in that if I remove the actual fix and run the new test, the test does fail— it detects inequality of the events on line 282. However, that is because of an implementation detail that I'd rather not rely on: the use of the `streams.deferredEvent` type, which memoizes the value of the event. That forces Go to treat the equality test as reference equality because the type contains things that can't be compared by value. But if the concrete types of `event1` and `event2` were a simpler struct type that just had fields for the event name and data, then line 282 would not work. I'd rather have this test be independent of such implementation details that aren't directly relevant to the logic we're testing, so I still think it's best to use an instrumented store as I described.